modules/network/src/main/java/org/apache/ignite/internal/network/recovery/RecoveryServerHandshakeManager.java

```diff
@@ -231,7 +232,12 @@ private void handleStaleClientId(HandshakeStartResponseMessage msg) {
                 .message(message)
                 .build();
 
-        sendHandshakeRejectedMessage(rejectionMessage, message);
+        ChannelFuture sendFuture = channel.writeAndFlush(new OutNetworkObject(rejectionMessage, emptyList(), false));
+
+        NettyUtils.toCompletableFuture(sendFuture).whenComplete((unused, throwable) -> {
+            // Ignoring ex as the most important thing here is to complete the future with the correct exception.
+            handshakeCompleteFuture.completeExceptionally(new HandshakeException(message));
+        });
```

Comment:
This is also very similar to sections of code in the neighboring class, maybe we can select a separate auxiliary method so as not to duplicate it?